[Pseudo-IPv4] Fix client IP resolution in addProxyIntegrationHeaders INTER-1139

erayaydin opened the pull request:
This PR fixes how the client IP is determined in `addProxyIntegrationHeaders`. Previously, `FPJS-Proxy-Client-IP` could be set incorrectly when Cloudflare's [Pseudo IPv4][cf-pseudo-ipv4] feature was enabled.

## What's Changed?

- Now correctly prioritizes `CF-Connecting-IPv6` when `Cf-Pseudo-IPv4` matches `CF-Connecting-IP`.
- Added an early return when `proxySecret` isn't set.
- Implemented a new `getIPFromHeaders` function to handle IP resolution.
- Added unit tests.

## Why This Matters?

This fix ensures that when Cloudflare's `Overwrite Headers` option is enabled, the correct client IP is used.

## How To Test?

- Run `pnpm test` to verify all unit tests pass.
- Try different header combinations (`CF-Connecting-IP`, `CF-Connecting-IPv6`, `Cf-Pseudo-IPv4`) and confirm expected result.

Let me know if anything needs adjusting! 😇 

[cf-pseudo-ipv4]: https://developers.cloudflare.com/network/pseudo-ipv4/

erayaydin committed:
fix: correctly resolve client IP from headers

Fixed incorrect logic in `addProxyIntegrationHeaders` by using
`getIPFromHeaders` to resolve the client IP. Now,
`FPJS-Proxy-Client-IP` correctly prioritizes `CF-Connecting-IPv6`
when `Cf-Pseudo-IPv4` matches `CF-Connecting-IP`, ensuring proper
handling of CloudFlare's [Pseudo IPv4][cf-pseudo-ipv4] feature when
the `Overwrite Headers` option is enabled. Also added missing
_early return_ when `proxySecret` is not set. Added unit tests for
`getIPFromHeaders` method.

[cf-pseudo-ipv4]: https://developers.cloudflare.com/network/pseudo-ipv4/

Related-Task: INTER-1139

erayaydin committed:
test: add test for empty CF-Connecting-IP header

Added a test case to check that the function returns an empty
string when the `CF-Connecting-IP` header is set to an empty
string.

Related-Task: INTER-1139

erayaydin committed:
refactor: expose getIPFromHeaders in the utils exports

This commit updates the exports in `src/utils/index.ts` to include
the `getIPFromHeaders` function.

Related-Task: INTER-1139

necipallef commented on test/utils/addProxyIntegrationHeaders.test.ts:
```diff
@@ -40,3 +40,43 @@ describe('addProxyIntegrationHeaders', () => {
+    expect(getIPFromHeaders(headers)).toEqual(ipv6)
+  })
+
+  it('returns CF-Connecting-IPv6 when Cf-Pseudo-IPv4 matches CF-Connecting-IP', () => {
```
I thought we would be adding test cases for `addProxyIntegrationHeaders`, this approach works, too. But I strongly insist to add at least one more test case for `addProxyIntegrationHeaders`, which would be this exact case, 'returns CF-Connecting-IPv6 when Cf-Pseudo-IPv4 matches CF-Connecting-IP'.

erayaydin replied:
Hey @necipallef I added new test cases and refactored `addProxyIntegrationHeaders` tests with commit [#64a88f0b](https://github.com/fingerprintjs/fingerprintjs-pro-cloudflare-worker/pull/208/commits/64a88f0b1c6ebc2c8fe976658d5d44531ef912c0).

erayaydin committed:
test: update addProxyIntegrationHeaders tests to handle IPv4 and IPv6 cases

Refactored tests to use variables for IPv4 and IPv6 values. Updated
tests to check if proxy headers are appended correctly. Added
additional tests for scenarios with Pseudo-IPv4 headers.

JuroUhlar commented on src/utils/addProxyIntegrationHeaders.ts:
```diff
@@ -2,9 +2,21 @@ import { getProxySecret, WorkerEnv } from '../env'
+export function getIPFromHeaders(headers: Headers) {
+  const connectingIP = headers.get('CF-Connecting-IP')
+
+  if (headers.get('Cf-Pseudo-IPv4') === connectingIP) {
```
I would add an explaining comment here

```suggestion
  // Correctly handle Cloudflare's [Pseudo IPv4](https://developers.cloudflare.com/network/pseudo-ipv4/) feature set to `Overwrite headers`
  if (headers.get('Cf-Pseudo-IPv4') === connectingIP) {
```

erayaydin replied:
Hey @JuroUhlar thanks for the suggestion. I added the comment with commit [#436e06b6](https://github.com/fingerprintjs/fingerprintjs-pro-cloudflare-worker/pull/208/commits/436e06b6eb5b8493521af6053f403c09137d5d0b) and also added you to `Co-Authored-By`. Could you please check and resolve the conversation if it's fixed?

erayaydin committed:
chore: add comment about handling CloudFlare's Pseudo IPv4 feature in getIPFromHeaders

Added a comment to the `getIPFromHeaders` function to explain the
handling of CloudFlare's Pseudo IPv4 feature.

Co-Authored-By: @JuroUhlar
